React/React hooks.md

```diff
@@ -0,0 +1,386 @@
+  }, [propA, propB]);
+
+  const memoizedArray = useMemo(() => [propA, propB], [propA, propB])
+
+  return (
+    <ChildComponent objectProp={memoizedObject} arrayProp={memoizedArray}/>
+  );
+}
+```
+
+#### Do not use `useMemo` as a garantee that it will be a constant throughout component rerenders
+
+If you need a value to stay the same throughout rerenders, you might think of `useMemo` as a nice way to "mimick" a constant. While it might seem that way, [it is not garanteed](https://reactjs.org/docs/hooks-reference.html#usememo) that it will not get garbage collected and that the value will not be recalculated.
+
+```jsx
+export const MyComponent: FC = () => {
+  const id = useMemo(() => generateId(), []); // Not garanteed to be a constant
+
+  // ...
+}
+```
+
+Alternative approach, with lazy initialization:
+
+```jsx
+export const MyComponent: FC = () => {
+  const [id, _] = useState(() => generateId()); // Initial value will stay the same throughout rerenders
+
+  // ...
+}
+```
+
+You can also store the value in a ref:
+_Note: updating ref values will not trigger a rerender._
+
+```jsx
+export const MyComponent: FC = () => {
+  const id = useRef(null);
+
+  if (!id.current) {
+    id.current = generateId();
+  }
+
+  // ...
+}
+```
+
+If you want the component to rerender when the value is set:
+
+```jsx
+export const Component: FC = () => {
+  const [id, setId] = useState();
+
+  useEffect() => {
+    if (!id) {
+      setId(generateId());
+    }
+  }, [id]);
+
+  // ...
+}
+```
+
+#### Wrap functions with `useCallback` it they can cause large subtrees to rerender too often
+
+While it is mostly unnecessary to wrap functions with `useCallback`, since the memoization process will usually be as expensive or more, it can be useful if functions' dependencies change often or if they are passed down to a large subtree of children.
+
+```jsx
+// In this case, memoization is probably not worth it, since it will not necessarily improve performance
+export const MyButton: FC = () => {
+  const onClick = useCallback((e: SyntheticEvent<HTMLButtonElement>) => {
+    ...
+  }, [])
+
+  return (
+    <Button onClick={onClick} />
+  );
+}
+```
+
+```jsx
+// Here, memoization is probably useful, because `onClick` will stay the same if other state rerenders the component, and will not trigger a large subtree to rerender unnecessarily
+export const MyList: FC = () => {
+  // ...
+
+  const onListItemClick = useCallback(() => {
+    ...
+  }, [...])
+
+  return <LargeList onListItemClick={onListItemClick} />;
+}
+```
+
+#### You don't have to avoid "inlining" non-expensive functions
+
+React is good at optimizing, so if you prematurly decide to wrap a function inside a `useCallback` and declare it in the component scope, you might get slower performance in many cases, because how memoization can be expensive. In most cases, even if you improve performance, it will be insignificant in comparison to making your code less clean and harder to mantain.
```

Comment:
The first sentence seems weird and too long. Maybe split it into two and start the second one with "Because memoization can be expensive"

Reply:
@DarkoKukovec I think it's better now